File: scripts/apply-migration.js
/**
 * Script pour appliquer la migration SQL à Supabase
 *
 * DEUX MÉTHODES:
 *
 * MÉTHODE 1 (Manuelle - recommandée):
 * 1. Allez sur https://supabase.com/dashboard/project/dnnwwexpmqhncnaizbfy/sql
 * 2. Copiez le contenu de scripts/migration-to-run.sql
 * 3. Collez dans l'éditeur SQL
 * 4. Cliquez sur "Run"
 *
 * MÉTHODE 2 (Automatique avec ce script):
 * 1. Récupérez votre mot de passe PostgreSQL:
 *    - Dashboard > Settings > Database > Connection String
 *    - Copiez le mot de passe après "postgresql://postgres:"
 * 2. Lancez: node scripts/apply-migration.js
 * 3. Entrez le mot de passe quand demandé
 */

const { Client } = require('pg');
const fs = require('fs');
const path = require('path');
require('dotenv').config({ path: '.env.local' });

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;

if (!SUPABASE_URL) {
  console.error('❌ NEXT_PUBLIC_SUPABASE_URL non trouvé dans .env.local');
  process.exit(1);
}

// Extraire le project reference
const match = SUPABASE_URL.match(/https:\/\/([^.]+)\.supabase\.co/);
if (!match) {
  console.error('❌ URL Supabase invalide');
  process.exit(1);
}
const projectRef = match[1];

async function main() {
  console.log('\n📤 Application de la migration SQL à Supabase');
  console.log('📍 Projet:', projectRef);
  console.log('\nVous avez DEUX options:\n');
  console.log('  OPTION 1 (Manuelle - Recommandée):');
  console.log('  1. Allez sur: https://supabase.com/dashboard/project/' + projectRef + '/sql');
  console.log('  2. Ouvrez le fichier: scripts/migration-to-run.sql');
  console.log('  3. Copiez-collez le contenu dans l\'éditeur SQL');
  console.log('  4. Cliquez sur "Run"\n');
  console.log('  OPTION 2 (Automatique):');
  console.log('  Vous aurez besoin de votre mot de passe PostgreSQL');
  console.log('  Dashboard > Settings > Database > Connection String > Copy\n');

  const readline = require('readline');
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  const choice = await new Promise(resolve => {
    rl.question('Quelle option choisissez-vous ? (1/2) [1]: ', ans => resolve(ans || '1'));
  });
  rl.close();

  if (choice === '1') {
    console.log('\n✅ Utilisez le fichier scripts/migration-to-run.sql');
    console.log('🔗 Lien direct: https://supabase.com/dashboard/project/' + projectRef + '/sql\n');
    return;
  }

  // Option 2: Automatique
  const rl2 = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  const dbPassword = await new Promise(resolve => {
    rl2.question('\nEntrez le mot de passe PostgreSQL: ', resolve);
  });
  rl2.close();

  if (!dbPassword) {
    console.log('❌ Mot de passe requis. Utilisez l\'option 1.');
    return;
  }

  const connectionString = `postgresql://postgres:${dbPassword}@db.${projectRef}.supabase.co:5432/postgres?sslmode=require`;

  console.log('\n🔗 Connexion à la base de données...');

  const client = new Client({ connectionString });

  try {
    await client.connect();
    console.log('✅ Connecté!\n');

    // Lire le fichier de migration
    const migrationPath = path.join(__dirname, '../supabase/migrations/001_initial_schema.sql');
    const migrationSQL = fs.readFileSync(migrationPath, 'utf8');

    console.log('📜 Exécution de la migration... (cela peut prendre quelques secondes)\n');

    // Exécuter la migration complète
    await client.query(migrationSQL);

    console.log('✅ Migration appliquée avec succès!\n');

    // Vérifier les tables
    console.log('📋 Tables créées:\n');
    const tables = [
      'profiles',
      'zones',
      'zone_payment_methods',
      'tarifs',
      'tickets',
      'pending_requests',
      'transactions',
      'sms_webhook_tokens',
      'ussd_templates',
    ];

    let allGood = true;
    for (const table of tables) {
      const result = await client.query(`
        SELECT EXISTS (
          SELECT FROM information_schema.tables
          WHERE table_schema = 'public'
          AND table_name = $1
        );
      `, [table]);

      if (result.rows[0].exists) {
        console.log(`   ✅ ${table}`);
      } else {
        console.log(`   ❌ ${table} - MANQUANTE`);
        allGood = false;
      }
    }

    if (allGood) {
      console.log('\n✨ Migration terminée avec succès!');
      console.log('   Toutes les tables sont créées.\n');
    } else {
      console.log('\n⚠️  Certaines tables manquent. Vérifiez les erreurs ci-dessus.\n');
    }

  } catch (err) {
    console.error('\n❌ Erreur:', err.message);
    if (err.message.includes('password authentication failed')) {
      console.error('   Le mot de passe PostgreSQL est incorrect.');
      console.error('   Récupérez-le depuis: Dashboard > Settings > Database > Connection String');
    }
    process.exit(1);
  } finally {
    await client.end();
  }
}

main().catch(err => {
  console.error('Erreur:', err);
  process.exit(1);
});

File: scripts/apply-migration.py
#!/usr/bin/env python3
"""
Script pour appliquer la migration SQL à Supabase

Usage: python scripts/apply-migration.py

Ce script va:
1. Demander le mot de passe PostgreSQL
2. Se connecter à Supabase
3. Appliquer la migration SQL complète
4. Vérifier que toutes les tables sont créées
"""

import os
import sys
from pathlib import Path
from dotenv import load_dotenv

try:
    import psycopg
    from psycopg.rows import dict_row
except ImportError:
    print("❌ psycopg3 n'est pas installé")
    print("Installez-le avec: pip install psycopg[binary] python-dotenv")
    sys.exit(1)

# Charger les variables d'environnement
load_dotenv('.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
if not SUPABASE_URL:
    print("❌ NEXT_PUBLIC_SUPABASE_URL non trouvé dans .env.local")
    sys.exit(1)

# Extraire le project reference
match = SUPABASE_URL.split('https://')[1].split('.supabase.co')[0]
project_ref = match

print(f"\n📤 Application de la migration SQL à Supabase")
print(f"📍 Projet: {project_ref}")
print(f"\nVous aurez besoin de votre mot de passe PostgreSQL.")
print(f"Récupérez-le depuis: https://supabase.com/dashboard/project/{project_ref}/settings/database\n")

# Demander le mot de passe
db_password = input("Entrez le mot de passe PostgreSQL: ").strip()
if not db_password:
    print("❌ Mot de passe requis")
    sys.exit(1)

# Chaîne de connexion
conn_string = f"postgresql://postgres:{db_password}@db.{project_ref}.supabase.co:5432/postgres?sslmode=require"

print("\n🔗 Connexion à la base de données...")

try:
    # Connexion
    conn = psycopg.connect(conn_string, autocommit=False)
    conn.row_factory = dict_row
    cursor = conn.cursor()

    print("✅ Connecté!\n")

    # Lire le fichier de migration
    migration_path = Path(__file__).parent.parent / 'supabase' / 'migrations' / '001_initial_schema.sql'
    with open(migration_path, 'r', encoding='utf-8') as f:
        migration_sql = f.read()

    print("📜 Exécution de la migration... (peut prendre quelques secondes)\n")

    # Exécuter la migration
    cursor.execute(migration_sql)
    conn.commit()

    print("✅ Migration appliquée avec succès!\n")

    # Vérifier les tables
    print("📋 Vérification des tables créées:\n")

    tables = [
        'profiles',
        'zones',
        'zone_payment_methods',
        'tarifs',
        'tickets',
        'pending_requests',
        'transactions',
        'sms_webhook_tokens',
        'ussd_templates',
    ]

    cursor.execute("""
        SELECT table_name
        FROM information_schema.tables
        WHERE table_schema = 'public'
        AND table_type = 'BASE TABLE'
    """)

    existing_tables = {row['table_name'] for row in cursor.fetchall()}

    all_good = True
    for table in tables:
        if table in existing_tables:
            print(f"   ✅ {table}")
        else:
            print(f"   ❌ {table} - MANQUANTE")
            all_good = False

    cursor.close()
    conn.close()

    if all_good:
        print("\n✨ Migration terminée avec succès!")
        print("   Toutes les tables sont créées.\n")
    else:
        print("\n⚠️  Certaines tables manquent.")
        print("   Vérifiez les erreurs ci-dessus.\n")

except Exception as err:
    print(f"\n❌ Erreur: {err}")
    if "password authentication failed" in str(err):
        print("   Le mot de passe PostgreSQL est incorrect.")
        print(f"   Récupérez-le depuis: https://supabase.com/dashboard/project/{project_ref}/settings/database")
    sys.exit(1)
